### ISSUE TITLE
False-positive `call-non-callable` from SymPy functions

### ISSUE THREAD

**Issue by augustfe** [astral-sh/ty#2825]
### Summary

When using `sympy.Function("f")`, ty reveals the type as `sympy.core.function.Function`, and then reports `call-non-callable` for `f(x)`. Instantiating a Function this way actually produces a `sympy.core.function.UndefinedFunction`, which is callable.

Reproduction, casting `f` to `UndefinedFunction` removes the issue.
```python
import sympy as sp

x = sp.symbols("x")
f = sp.Function("f")
f(x)  # error[call-non-callable]
```

Diagnostic result from `uv run ty check main.py`:
```system
error[call-non-callable]: Object of type `Function` is not callable
 --> main.py:7:1
  |
5 | # Actual: sympy.core.function.UndefinedFunction
6 | f = sp.Function("f")
7 | f(x)
  | ^^^^
  |
info: rule `call-non-callable` is enabled by default

Found 1 diagnostic
```

The issue likely stems from [`Function.__new__`](https://github.com/sympy/sympy/blob/16fa855354eb7bcabd3fe10993841e03b1382692/sympy/core/function.py#L444-L450), a common pattern in SymPy:
```python
class Function(Application, Expr):

    @cacheit
    def __new__(cls, *args, **options) -> type[AppliedUndef]:  # type: ignore
        # Handle calls like Function('f')
        if cls is Function:
            return UndefinedFunction(*args, **options)  # type: ignore
        else:
            return cls._new_(*args, **options)  # type: ignore
```

### Version

ty 0.0.17 (8cec85718 2026-02-13)

**Comment by sharkdp** [comment]
Thank you for reporting this.

> The issue likely stems from [`Function.__new__`](https://github.com/sympy/sympy/blob/16fa855354eb7bcabd3fe10993841e03b1382692/sympy/core/function.py#L444-L450), a common pattern in SymPy:

Yes, it seems likely that the root cause is #281

**Comment by carljm** [comment]
#281 is now fixed, but the remaining problem here is the `@cacheit` decorator, which is untyped, so it relies on typecheckers optimistically assuming that untyped decorators pass through the decorated function unchanged. So we still don't respect the return type of this `__new__` method, pending #1971.
